packages/flutter_tools/lib/src/build_system/targets/assets.dart

```diff
@@ -112,6 +114,56 @@ Future<Depfile> copyAssets(Environment environment, Directory outputDirectory, {
         resource.release();
       }
   }));
+
+  // Copy deferred components assets only for release or profile builds.
```

Comment:
This seems like its repeating a lot of the logic of copyAssets. Consider:

1. configure the output path and deferredComponents instead of looking them up on the project. (i.e. `deferredComponents: true, componentOutputDir: ...`. You can let the build rule set it to false on build/profile modes et cetera